File: src/lexer.ts
import { Token, TokenKind, createToken, isDigit, isLetter, isWhitespace, lookupIdent } from './token';

export class Lexer {
  private input: string;
  private position: number;
  private readPosition: number;
  private ch: string;

  constructor(input: string) {
    this.input = input;
    this.position = 0;
    this.readPosition = 0;
    this.ch = '';
    this.readChar();
  }

  /** @throws {Error} */
  public nextToken(): Token {
    let token: Token;

    this.consumeWhitespace();

    switch (this.ch) {
      case '=':
        if (this.peekChar() === TokenKind.Assign) {
          this.readChar();
          token = createToken(TokenKind.Equal, TokenKind.Equal);
        } else {
          token = createToken(TokenKind.Assign, this.ch);
        }
        break;
      case '!':
        if (this.peekChar() === TokenKind.Assign) {
          this.readChar();
          token = createToken(TokenKind.NotEqual, TokenKind.NotEqual);
        } else {
          token = createToken(TokenKind.Bang, this.ch);
        }
        break;
      case ';':
        token = createToken(TokenKind.Semicolon, this.ch);
        break;
      case '(':
        token = createToken(TokenKind.LParen, this.ch);
        break;
      case ')':
        token = createToken(TokenKind.RParen, this.ch);
        break;
      case ',':
        token = createToken(TokenKind.Comma, this.ch);
        break;
      case '+':
        token = createToken(TokenKind.Plus, this.ch);
        break;
      case '-':
        token = createToken(TokenKind.Minus, this.ch);
        break;
      case '/':
        token = createToken(TokenKind.Slash, this.ch);
        break;
      case '*':
        token = createToken(TokenKind.Asterisk, this.ch);
        break;
      case '<':
        token = createToken(TokenKind.LT, this.ch);
        break;
      case '>':
        token = createToken(TokenKind.GT, this.ch);
        break;
      case '{':
        token = createToken(TokenKind.LBrace, this.ch);
        break;
      case '}':
        token = createToken(TokenKind.RBrace, this.ch);
        break;
      case '\0':
        token = createToken(TokenKind.EOF, '');
        break;
      default:
        if (isLetter(this.ch)) {
          const literal = this.readIdentifier();
          const type = lookupIdent(literal as keyof typeof TokenKind);
          return createToken(type, literal);
        } else if (isDigit(this.ch)) {
          const literal = this.readNumber();
          return createToken(TokenKind.Int, literal);
        } else {
          return createToken(TokenKind.Illegal, this.ch);
        }
    }

    this.readChar();

    return token;
  }

  private readIdentifier(): string {
    const startPosition = this.position;

    while (isLetter(this.ch)) {
      this.readChar();
    }

    return this.input.substring(startPosition, this.position);
  }

  private readNumber(): string {
    const startPosition = this.position;
    while (isDigit(this.ch)) {
      this.readChar();
    }
    return this.input.substring(startPosition, this.position);
  }

  private readChar(): void {
    if (this.readPosition >= this.input.length) {
      this.ch = '\0';
    } else {
      this.ch = this.input[this.readPosition];
    }

    this.position = this.readPosition;
    this.readPosition++;
  }

  private peekChar(): string {
    if (this.readPosition >= this.input.length) {
      return '\0';
    } else {
      return this.input[this.readPosition];
    }
  }

  private consumeWhitespace(): void {
    while (isWhitespace(this.ch)) {
      this.readChar();
    }
  }
}

File: lexer/lexer.go
package lexer

import (
	"github.com/marianozunino/monkey-lang/token"
)

type Lexer struct {
	input      string
	currentPos int  // current position in input
	readPos    int  // current reading position in input
	ch         byte // current char
}

func New(input string) *Lexer {
	l := &Lexer{input: input}
	l.readChar() // initialize currentPos, readPos and ch
	return l
}

func (l *Lexer) readChar() {
	if l.readPos >= len(l.input) {
		l.ch = 0 // EOF?
	} else {
		l.ch = l.input[l.readPos]
		l.currentPos = l.readPos
		l.readPos++
	}
}

func (l *Lexer) NextToken() token.Token {
	l.consumeWhitespace()

	var tkn token.Token

	switch string(l.ch) {

	case token.COMMA,
		token.SEMICOLON,

		token.LPAREN,
		token.RPAREN,
		token.LBRACE,
		token.RBRACE,

		token.PLUS,
		token.MINUS,
		token.ASTERISK,
		token.SLASH,

		token.GT,
		token.LT,

		token.EOF:

		tkn = newToken(token.TokenType(l.ch), string(l.ch))
	case token.BANG:
		if l.peekChar() != '=' {
			tkn = newToken(token.TokenType(l.ch), string(l.ch))
		} else {
			l.readChar()
			tkn = newToken(token.TokenType(token.NOT_EQ), token.NOT_EQ)
		}
	case token.ASSIGN:
		if l.peekChar() == '=' {
			l.readChar()
			tkn = newToken(token.TokenType(token.EQ), token.EQ)
		} else {
			tkn = newToken(token.TokenType(l.ch), string(l.ch))
		}
	default:
		if isLetter(l.ch) {
			ident := l.readIdentifier()
			return newToken(token.LookupIdent(ident), ident)
		} else if isDigit(l.ch) {
			number := l.readNumber()
			return newToken(token.INT, number)
		} else {
			tkn = newToken(token.ILLEGAL, string(token.ILLEGAL))
		}
	}

	l.readChar()

	return tkn
}

func isLetter(ch byte) bool {
	return ('A' <= ch && ch <= 'z') || ch == '_'
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func (l *Lexer) peekChar() byte {
	if l.readPos >= len(l.input) {
		return '\x00'
	} else {
		return l.input[l.readPos]
	}
}

func (l *Lexer) readNumber() string {
	position := l.currentPos

	for isDigit(l.ch) {
		l.readChar()
	}

	return l.input[position:l.currentPos]
}

func (l *Lexer) readIdentifier() string {
	position := l.currentPos

	for isLetter(l.ch) {
		l.readChar()
	}

	return l.input[position:l.currentPos]
}

func (l *Lexer) consumeWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
		l.readChar()
	}
}

func newToken(tknType token.TokenType, literal string) token.Token {
	return token.Token{
		Type:    tknType,
		Literal: literal,
	}
}
